In [5]:
import cohere
import numpy as np

In [45]:
co = cohere.ClientV2(api_key="CVjpkOY6TMQ3VIfYqBVrBiYhoninecium8PeGMxZ")

In [7]:
products = [
  {"id": 1, "name": "Classic Leather Sneakers", "category": "footwear", "price": 89.99, "color": "white", "description": "Minimalist leather sneakers with rubber sole, everyday wear"},
  {"id": 2, "name": "Running Shoes Pro", "category": "footwear", "price": 129.99, "color": "black", "description": "Lightweight mesh running shoes with cushioned midsole"},
  {"id": 3, "name": "Canvas High-Top Sneakers", "category": "footwear", "price": 64.99, "color": "red", "description": "Casual canvas sneakers, high-top design, vintage style"},
  {"id": 4, "name": "Wireless Noise-Cancelling Headphones", "category": "electronics", "price": 249.99, "color": "black", "description": "Over-ear headphones with active noise cancellation, 30hr battery"},
  {"id": 5, "name": "True Wireless Earbuds", "category": "electronics", "price": 99.99, "color": "white", "description": "Compact earbuds with charging case, sweat resistant"},
  {"id": 6, "name": "Mechanical Keyboard RGB", "category": "electronics", "price": 119.99, "color": "black", "description": "Tactile mechanical keyboard with customizable RGB backlight"},
  {"id": 7, "name": "Stainless Steel Water Bottle", "category": "home", "price": 24.99, "color": "silver", "description": "Insulated bottle keeps drinks cold 24h, 750ml capacity"},
  {"id": 8, "name": "Ceramic Coffee Mug Set", "category": "home", "price": 19.99, "color": "blue", "description": "Set of 4 minimalist ceramic mugs, dishwasher safe"},
  {"id": 9, "name": "Cast Iron Skillet", "category": "home", "price": 44.99, "color": "black", "description": "Pre-seasoned cast iron skillet, oven safe, 12-inch"},
  {"id": 10, "name": "Merino Wool Sweater", "category": "apparel", "price": 79.99, "color": "gray", "description": "Soft merino wool crewneck sweater, breathable warmth"},
  {"id": 11, "name": "Denim Jacket", "category": "apparel", "price": 69.99, "color": "blue", "description": "Classic fit denim jacket with button closure"},
  {"id": 12, "name": "Yoga Mat Premium", "category": "fitness", "price": 39.99, "color": "purple", "description": "Non-slip yoga mat, 6mm thick, eco-friendly material"},
  {"id": 13, "name": "Adjustable Dumbbell Set", "category": "fitness", "price": 159.99, "color": "black", "description": "Adjustable weight dumbbells, 5-25kg range, space saving"},
  {"id": 14, "name": "Smart Fitness Watch", "category": "electronics", "price": 199.99, "color": "black", "description": "Tracks heart rate, sleep, and workouts, 7-day battery"},
  {"id": 15, "name": "Leather Crossbody Bag", "category": "apparel", "price": 89.99, "color": "brown", "description": "Genuine leather crossbody bag with adjustable strap"}
]

In [9]:
def product_to_text(product: dict) -> str:
    """Build a single searchable string for each product"""
    return (
        f"Name: {product['name']} - "
        f"Category: {product['category']} - "
        f"Price: {product['price']} - "
        f"Color: {product['color']} - "
        f"Description: {product['description']}"
    )

In [13]:
texts = [product_to_text(product) for product in products]
texts

['Name: Classic Leather Sneakers - Category: footwear - Price: 89.99 - Color: white - Description: Minimalist leather sneakers with rubber sole, everyday wear',
 'Name: Running Shoes Pro - Category: footwear - Price: 129.99 - Color: black - Description: Lightweight mesh running shoes with cushioned midsole',
 'Name: Canvas High-Top Sneakers - Category: footwear - Price: 64.99 - Color: red - Description: Casual canvas sneakers, high-top design, vintage style',
 'Name: Wireless Noise-Cancelling Headphones - Category: electronics - Price: 249.99 - Color: black - Description: Over-ear headphones with active noise cancellation, 30hr battery',
 'Name: True Wireless Earbuds - Category: electronics - Price: 99.99 - Color: white - Description: Compact earbuds with charging case, sweat resistant',
 'Name: Mechanical Keyboard RGB - Category: electronics - Price: 119.99 - Color: black - Description: Tactile mechanical keyboard with customizable RGB backlight',
 'Name: Stainless Steel Water Bottle 

In [32]:
model = "embed-v4.0"
input_type = "search_document"

In [33]:
response = co.embed(
    model=model,
    input_type=input_type,
    texts=texts,
    embedding_types=["float"],
    output_dimension=1024,
)

In [34]:
embeddings = response.embeddings.float
embeddings

[[-0.05215687,
  -0.05838135,
  -0.09615752,
  0.0008183048,
  -0.05129832,
  -0.049366582,
  -0.042712826,
  0.011697734,
  -0.059669178,
  -0.025005247,
  0.026829664,
  0.051512957,
  0.05022513,
  0.032624874,
  0.024468655,
  0.023717424,
  0.021893006,
  -0.02232228,
  -0.015024613,
  0.027258938,
  0.0011469682,
  0.028546765,
  0.042283554,
  0.0067610755,
  -0.008585493,
  0.013253855,
  -0.0018646618,
  0.03176632,
  -0.01588316,
  0.008853789,
  -0.0045342133,
  0.024683291,
  -0.0026963814,
  -0.022644237,
  -0.017170986,
  0.045288473,
  -0.045288473,
  -0.021141777,
  0.029405314,
  -0.02403938,
  0.042712826,
  -0.026078435,
  -0.026722346,
  0.010570888,
  -0.03176632,
  -0.009712339,
  -0.008639152,
  -0.0005265322,
  0.0149172945,
  -0.027473576,
  -0.04013718,
  0.042712826,
  -0.03133705,
  0.004158598,
  0.037561532,
  -0.036488343,
  0.04550311,
  -0.0020390546,
  -0.016312437,
  0.00082501216,
  0.0070293727,
  -0.005848867,
  -0.02564916,
  0.0055537405,
  -0.04

In [35]:
embedded_products = [
    {
        "id": product["id"],
        "text": text,
        "embedding": vector,
        **product,
    }
    for product, text, vector in zip(products, texts, embeddings)
]

In [36]:
print(len(embedded_products), len(embedded_products[0]["embedding"]))


15 1024


In [37]:
print(embedded_products[0]["text"][:80], "...")


Name: Classic Leather Sneakers - Category: footwear - Price: 89.99 - Color: whit ...


In [38]:
def cosine_similarity(a, b) -> float:
    """Cosine similarity between two embedding vectors."""
    a, b = np.array(a), np.array(b)
    return float(np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b)))


In [39]:
# np.linalg -> numpy linear algebra
# np.dot -> dot product
# np.linalg.norm -> norm -> magnitude of the vector
# np.array -> array
# np.float -> float
# np.float64 -> float64
# np.float32 -> float32

In [40]:
query = "comfortable shoes for running"
query_emb = co.embed(
    model=model,
    input_type="search_query",
    texts=[query],
    embedding_types=["float"],
    output_dimension=1024,
).embeddings.float[0]

In [41]:
query_emb

[0.056013376,
 -0.032428797,
 0.01835845,
 -0.00050669996,
 -0.039128963,
 -0.010452257,
 -0.015812388,
 0.02371858,
 0.009514233,
 -0.022646556,
 -0.0027805683,
 0.114170805,
 0.04154102,
 -0.078257926,
 0.0146063585,
 0.00081239495,
 0.027470674,
 0.035644874,
 -0.02693466,
 0.07128975,
 0.02278056,
 -0.05815743,
 -0.041005004,
 0.044489093,
 -0.011658287,
 0.020234497,
 -0.03537687,
 0.024924612,
 0.030284744,
 -0.05333331,
 -0.014740361,
 -0.03216079,
 -0.024924612,
 0.0070351725,
 -0.0078056916,
 0.008710214,
 -0.008040197,
 0.0012060297,
 0.05654939,
 0.0100502465,
 0.054405335,
 0.039664973,
 -0.051725272,
 -0.0002994136,
 -0.016884414,
 0.004723616,
 0.003098826,
 0.034036838,
 0.023852587,
 0.048777197,
 0.0071021747,
 -0.018894464,
 -0.05654939,
 -0.01346733,
 0.030284744,
 -0.056817397,
 -0.0050251232,
 -0.033500824,
 0.0016080394,
 0.019296475,
 -0.008978221,
 -0.04368507,
 0.018090444,
 0.006700164,
 -0.072361775,
 0.01916247,
 -0.03859295,
 0.01996649,
 -0.05494135,
 0.02

In [42]:
scores = [
    (p["name"], cosine_similarity(query_emb, p["embedding"]))
    for p in embedded_products
]

In [43]:
sorted(scores, key=lambda x: x[1], reverse=True)[:3]


[('Running Shoes Pro', 0.432303558295639),
 ('Classic Leather Sneakers', 0.20489766895294018),
 ('True Wireless Earbuds', 0.19430687702153604)]

search_document  -> “this is something in my catalog/library”
search_query = “this is what someone is searching for”

In [3]:
import base64
import cohere
import numpy as np
import urllib.request

co = cohere.ClientV2(api_key="CVjpkOY6TMQ3VIfYqBVrBiYhoninecium8PeGMxZ")  # or os.environ["COHERE_API_KEY"]


def unsplash_to_data_url(url: str) -> str:
    """Download an Unsplash image and convert it to a Cohere data URL."""
    with urllib.request.urlopen(url) as response:
        image_bytes = response.read()
    content_type = response.headers.get_content_type()
    b64 = base64.b64encode(image_bytes).decode("utf-8")
    return f"data:{content_type};base64,{b64}"


def cosine_similarity(a, b) -> float:
    """Cosine similarity between two embedding vectors."""
    a, b = np.array(a), np.array(b)
    return float(np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b)))

In [13]:
products = [
    {
        "id": 1,
        "name": "Running Shoes Pro",
        "text": "Lightweight black running shoes with cushioned midsole",
        "unsplash_url": "https://images.unsplash.com/photo-1542291026-7eec264c27ff?w=800",
    },
    {
        "id": 2,
        "name": "Wireless Headphones",
        "text": "Over-ear headphones with active noise cancellation",
        "unsplash_url": "https://images.unsplash.com/photo-1505740420928-5e560c06d30e?w=800",
    },
    {
        "id": 3,
        "name": "Classic Leather Sneakers",
        "text": "Minimalist white leather sneakers for everyday wear",
        "unsplash_url": "https://images.unsplash.com/photo-1595950653106-6c9ebd614d3a?w=800",
    },
]

catalog_inputs = []
for product in products:
    image_data_url = unsplash_to_data_url(product["unsplash_url"])

    catalog_inputs.append(
        {
            "content": [
                {
                    "type": "text",
                    "text": f"{product['name']}. {product['text']}",
                },
                {
                    "type": "image_url",
                    "image_url": {"url": image_data_url},
                },
            ]
        }
    )

doc_response = co.embed(
    model="embed-v4.0",
    input_type="search_query",
    inputs=catalog_inputs,
    embedding_types=["float"],
)

doc_embeddings = doc_response.embeddings.float

In [14]:
query_image_url = "https://images.unsplash.com/photo-1595950653106-6c9ebd614d3a?w=800"
query_image_data_url = unsplash_to_data_url(query_image_url)

query_response = co.embed(
    model="embed-v4.0",
    input_type="search_query",   # important: query side
    inputs=[
        {
            "content": [
                {
                    "type": "image_url",
                    "image_url": {"url": query_image_data_url},
                }
            ]
        }
    ],
    embedding_types=["float"],
)

query_embedding = query_response.embeddings.float[0]

In [15]:
scores = [
    (product["name"], cosine_similarity(query_embedding, doc_embeddings[i]))
    for i, product in enumerate(products)
]

for name, score in sorted(scores, key=lambda x: x[1], reverse=True):
    print(f"{name}: {score:.4f}")

Classic Leather Sneakers: 0.3689
Running Shoes Pro: 0.1300
Wireless Headphones: 0.1240


In [18]:
import base64
import cohere
import numpy as np
import urllib.request

co = cohere.ClientV2(api_key="CVjpkOY6TMQ3VIfYqBVrBiYhoninecium8PeGMxZ")


def unsplash_to_data_url(url: str) -> str:
    """Download an Unsplash image and convert it to a Cohere data URL."""
    with urllib.request.urlopen(url) as response:
        image_bytes = response.read()
    content_type = response.headers.get_content_type()
    b64 = base64.b64encode(image_bytes).decode("utf-8")
    return f"data:{content_type};base64,{b64}"


def cosine_similarity(a, b) -> float:
    """Cosine similarity between two embedding vectors."""
    a, b = np.array(a), np.array(b)
    return float(np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b)))

In [19]:
products = [
    {
        "id": 1,
        "name": "Running Shoes Pro",
        "text": "Lightweight black running shoes with cushioned midsole",
        "unsplash_url": "https://images.unsplash.com/photo-1542291026-7eec264c27ff?w=800",
    },
    {
        "id": 2,
        "name": "Wireless Headphones",
        "text": "Over-ear headphones with active noise cancellation",
        "unsplash_url": "https://images.unsplash.com/photo-1505740420928-5e560c06d30e?w=800",
    },
    {
        "id": 3,
        "name": "Classic Leather Sneakers",
        "text": "Minimalist white leather sneakers for everyday wear",
        "unsplash_url": "https://images.unsplash.com/photo-1595950653106-6c9ebd614d3a?w=800",
    },
]

In [20]:
MODEL = "embed-v4.0"

text_inputs = [f"{p['name']}. {p['text']}" for p in products]

image_inputs = [
    {
        "content": [
            {
                "type": "image_url",
                "image_url": {"url": unsplash_to_data_url(p["unsplash_url"])},
            }
        ]
    }
    for p in products
]

text_embeddings = co.embed(
    model=MODEL,
    input_type="search_document",
    texts=text_inputs,
    embedding_types=["float"],
).embeddings.float

image_embeddings = co.embed(
    model=MODEL,
    input_type="search_document",
    inputs=image_inputs,
    embedding_types=["float"],
).embeddings.float

for i, product in enumerate(products):
    product["text_embedding"] = text_embeddings[i]
    product["image_embedding"] = image_embeddings[i]

In [21]:
def embed_text_query(query: str) -> list[float]:
    """Embed a text search query."""
    return co.embed(
        model=MODEL,
        input_type="search_query",
        texts=[query],
        embedding_types=["float"],
    ).embeddings.float[0]


def embed_image_query(image_url: str) -> list[float]:
    """Embed an image search query."""
    data_url = unsplash_to_data_url(image_url)
    return co.embed(
        model=MODEL,
        input_type="search_query",
        inputs=[
            {
                "content": [
                    {
                        "type": "image_url",
                        "image_url": {"url": data_url},
                    }
                ]
            }
        ],
        embedding_types=["float"],
    ).embeddings.float[0]


def rank_by_text(query: str, top_k: int = 3):
    """Search products using text only."""
    query_emb = embed_text_query(query)
    scores = [
        (p["name"], cosine_similarity(query_emb, p["text_embedding"]))
        for p in products
    ]
    return sorted(scores, key=lambda x: x[1], reverse=True)[:top_k]


def rank_by_image(image_url: str, top_k: int = 3):
    """Search products using image only."""
    query_emb = embed_image_query(image_url)
    scores = [
        (p["name"], cosine_similarity(query_emb, p["image_embedding"]))
        for p in products
    ]
    return sorted(scores, key=lambda x: x[1], reverse=True)[:top_k]


def rank_by_hybrid(
    query: str,
    image_url: str,
    text_weight: float = 0.5,
    image_weight: float = 0.5,
    top_k: int = 3,
):
    """Search products using text + image together."""
    text_query_emb = embed_text_query(query)
    image_query_emb = embed_image_query(image_url)

    scores = []
    for p in products:
        text_score = cosine_similarity(text_query_emb, p["text_embedding"])
        image_score = cosine_similarity(image_query_emb, p["image_embedding"])
        final_score = text_weight * text_score + image_weight * image_score
        scores.append((p["name"], final_score, text_score, image_score))

    return sorted(scores, key=lambda x: x[1], reverse=True)[:top_k]

In [22]:
rank_by_text("white minimalist sneakers")


[('Classic Leather Sneakers', 0.5731523987259316),
 ('Running Shoes Pro', 0.2708718897992223),
 ('Wireless Headphones', 0.1361357637593229)]

In [23]:
rank_by_image("https://images.unsplash.com/photo-1595950653106-6c9ebd614d3a?w=800")


[('Classic Leather Sneakers', 0.9850360058543738),
 ('Running Shoes Pro', 0.3370954366486602),
 ('Wireless Headphones', 0.2106242831314469)]

In [24]:
rank_by_hybrid(
    query="white everyday sneakers",
    image_url="https://images.unsplash.com/photo-1595950653106-6c9ebd614d3a?w=800",
    text_weight=0.5,
    image_weight=0.5,
)

[('Classic Leather Sneakers',
  0.7378409938643845,
  0.49064598187439523,
  0.9850360058543738),
 ('Running Shoes Pro',
  0.2679403688872393,
  0.19878530112581844,
  0.3370954366486602),
 ('Wireless Headphones',
  0.16854373160538633,
  0.12646318007932575,
  0.2106242831314469)]

# Resumen: tipos de vectores, búsquedas y aplicaciones

## 1. Tipos de representaciones

| Tipo | Forma | Qué captura | Almacenamiento |
|------|-------|-------------|----------------|
| **Dense (denso)** | Array de floats (`[0.12, -0.34, ...]`) | Significado semántico aprendido | Fijo (ej. 1024 × 4 bytes) |
| **Sparse (disperso)** | Pares `{índice: peso}` | Keywords + expansión lexical aprendida | Eficiente si pocos términos activos |
| **Full-text (BM25)** | Texto indexado (no es vector) | Coincidencia literal de tokens | Texto + índice invertido |
| **Cuantizado** | Dense comprimido (int8, binary) | Mismo significado, menos precisión | Menor memoria |
| **Multivector** | Varios vectores por documento/chunk | Precisión fina (ColBERT) | Alto coste |

### Gradiente de señales

```
Semántica (concepto)  ←————————————————————→  Coincidencia exacta (tokens)
     Dense                    Sparse aprendido              BM25 / Full-text
```

## 2. Tipos de búsqueda (Pinecone / RAG)

| Búsqueda | Señal | Cómo funciona | Fortaleza | Debilidad |
|----------|-------|---------------|-----------|-----------|
| **Semantic (dense)** | Embeddings | ANN + cosine/dotproduct | Sinónimos, paráfrasis, multimodal | Términos exactos, SKUs raros |
| **Full-text (BM25)** | Tokens literales | Índice invertido + BM25 | Nombres, frases, booleanos | No entiende sinónimos |
| **Sparse lexical** | Sparse aprendido | Producto interno en índices activos | Keywords inteligentes + expansión | Requiere encoder sparse |
| **Hybrid** | Dense + lexical | Combina scores (`alpha`) o filtro + ranking | Lo mejor de ambos mundos | Más complejidad, tuning |

### Filtro vs scoring

| | Filtro | Scoring |
|--|--------|---------|
| Naturaleza | Sí/No (determinista) | Ranking (probabilístico) |
| Ejemplos | `price < 50`, `$match_phrase` | BM25, cosine similarity |
| Cuándo | Antes de rankear | Ordena candidatos |

> **Nota:** precio, fecha, stock → **metadata/SQL**, no vectores.

## 3. Aplicación en búsqueda de productos

### Datos típicos por producto

```text
id, name, category, price, color, description
text_embedding   → dense (Cohere embed-v4.0)
image_embedding  → dense (opcional, multimodal)
texto crudo      → BM25 (opcional, nombres/SKUs exactos)
```

### Queries y herramienta recomendada

| Query del usuario | Herramienta | Por qué |
|-------------------|-------------|---------|
| `"zapatos cómodos para caminar"` | Dense | Intención semántica |
| `"Classic Leather Sneakers"` | BM25 / sparse | Nombre exacto |
| `"SKU-8842"` | BM25 + filtro | Identificador literal |
| Foto de zapatilla | `image_embedding` | Similitud visual |
| Texto + foto | Hybrid multimodal | Dos modalidades |
| `"los 5 más baratos"` | `ORDER BY price` | Orden numérico, no vector |
| `"electrónica < 100€"` | Filtro metadata | Condición dura |
| `"zapatillas baratas para correr"` | Dense → sort por `price` | Semántica + precio en app |

### Esquema recomendado por fase

| Fase | Qué guardar |
|------|-------------|
| **MVP** (este notebook) | 1 `text_embedding` por producto |
| **Producción básica** | Dense + texto para BM25 |
| **Multimodal** | `text_embedding` + `image_embedding` |
| **Producción completa** | Dense + BM25/sparse + metadata (`price`, `category`) |

### Lo que NO necesitas

- Varios densos del mismo texto con el mismo modelo
- Embeddings para ordenar por precio
- Chunking (los productos son documentos cortos)

## 4. Aplicación en documentos (RAG)

### Por extensión del documento

| Extensión | Ejemplo | Estrategia | Vector |
|-----------|---------|------------|--------|
| **Muy corto** | Título, FAQ, tweet | 1 doc = 1 vector | Dense o BM25 |
| **Corto–medio** | Párrafo, artículo | 1 doc = 1 vector | Dense |
| **Medio–largo** | Capítulo, manual | **Chunking** | Dense por chunk |
| **Muy largo** | PDF, libro | Chunks + metadata (`doc_id`, `page`) | Dense por chunk + hybrid opcional |

### Chunking (documentos largos)

```text
Documento largo
    ├─ chunk 1 (512 tokens) → vector₁
    ├─ chunk 2 (512 tokens) → vector₂
    └─ chunk 3 (512 tokens) → vector₃
```

| Parámetro | Valor habitual |
|-----------|----------------|
| `chunk_size` | 300–800 tokens |
| `overlap` | 10–20% |

### Queries y herramienta

| Query | Herramienta |
|-------|-------------|
| Pregunta en lenguaje natural | Dense por chunk |
| Término técnico / cláusula exacta | BM25 o sparse |
| Contrato: *"indemnización sección 4.2"* | Hybrid (dense + BM25) |
| Máxima precisión en corpus grande | Dense + reranker (o ColBERT) |

## 5. Hybrid search: patrones

| Patrón | Cuándo | Cómo combina |
|--------|--------|--------------|
| **Filtro + dense** | "Semántica pero solo docs con frase X" | `$match_phrase` + `dense_vector` |
| **Dense + sparse (mismo índice)** | Vector API, ambas señales | `alpha * dense + (1-alpha) * sparse` |
| **Dos búsquedas + merge** | Máximo control | Top-K dense + top-K BM25 → dedup → rerank |
| **Multimodal (productos)** | Texto + imagen | `rank_by_hybrid` con pesos por modalidad |

### Valores de `alpha` (dense vs sparse)

| Alpha | Sesgo | Caso de uso |
|-------|-------|-------------|
| `0.75` | Semántico | Lenguaje natural |
| `0.50` | Balanceado | Mix keyword + intención |
| `0.25` | Lexical | SKUs, IDs, nombres propios |

## 6. Árbol de decisión rápido

```text
¿Es búsqueda de productos o documentos?
│
├─ PRODUCTOS
│   ├─ ¿Query semántica? → Dense
│   ├─ ¿Nombre/SKU exacto? → BM25 / sparse
│   ├─ ¿Por imagen? → image_embedding
│   └─ ¿Por precio/stock? → Metadata (no vector)
│
└─ DOCUMENTOS
    ├─ ¿Cabe en límite del modelo? → 1 vector dense
    ├─ ¿Documento largo? → Chunking + dense por chunk
    ├─ ¿Términos exactos críticos? → + BM25/sparse por chunk
    └─ ¿Máxima precisión? → + reranker
```

## 7. Este proyecto

| Componente | Estado |
|------------|--------|
| Modelo | Cohere `embed-v4.0` |
| Vector | Dense float, 1024 dims |
| Productos | 15, documentos cortos |
| Búsqueda | `cosine_similarity` en memoria |
| Multimodal | `text_embedding` + `image_embedding` + `rank_by_hybrid` |

**Siguiente paso natural (si escala):** Pinecone con dense + BM25 sobre `name`/`description`, metadata `price`/`category` para filtros.

## 8. Cheat sheet final

| Necesitas… | Usa… | No uses… |
|------------|------|----------|
| Sinónimos / intención | Dense | Solo BM25 |
| Nombre exacto / SKU | BM25 / sparse | Solo dense |
| Ambos | Hybrid | Un solo tipo |
| Ordenar por precio | `ORDER BY price` | Embeddings |
| Doc largo | Chunks + dense | 1 vector del doc entero |
| Buscar por foto | `image_embedding` | Solo texto |
| Menos memoria | Cuantización | — |
| Máxima precisión docs | Reranker | Solo top-K dense |